### ISSUE TITLE
cannot build package with project references

### ISSUE THREAD

**Issue by dagda1** [ezolenko/rollup-plugin-typescript2#260]
#### What happens and why it is wrong

I  am trying to build in a package that has project references in the tsconfig.json

```json
{
  "extends": "../../tsconfig-base.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": "./src" 
  },
  "include": ["src/**/*.ts", "src/**/*.tsx"],
  "exclude": ["src/**/*.test.ts", "src/**/*.test.tsx"],
  "references": [
    { "path": "../util" }
  ]
}
```

The project builds fine if I run `yarn run tsc --build` but fails when using rollup and the `rollup-plugin-typescript2`. 

I can see in the source that if the tsconfig has project references then files from the references are included:

```ts
    if (parsedConfig.projectReferences) {
      included = lodash.concat(included, expandIncludeWithDirs(included, parsedConfig.projectReferences.map((x) => x.path)));
      excluded = lodash.concat(excluded, expandIncludeWithDirs(excluded, parsedConfig.projectReferences.map((x) => x.path)));
    }
```

Which will create an array of globs like this:

```ts
[
  '*.ts+(|x)',
  '**/*.ts+(|x)',
  '/Users/paulcowan/projects/cuttingedge/packages/util/*.ts+(|x)',
  '/Users/paulcowan/projects/cuttingedge/packages/util/**/*.ts+(|x)'
]
```

I then get this error:

> : semantic error TS6059: File '/Users/paulcowan/projects/cuttingedge/packages/util/src/index.ts' is not under 'rootDir' '/Users/paulcowan/projects/cuttingedge/packages/hooks'. 'rootDir' is expected to contain all source files.

Reading that error, I would then need to update my  tsconfig.json to this:

```json
{
  "extends": "../../tsconfig-base.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDirs": ["./src", "../util" ]
  },
  "include": ["src/**/*.ts", "src/**/*.tsx"],
  "exclude": ["./dist", "src/**/*.test.ts", "src/**/*.test.tsx"],
  "references": [
    { "path": "../util" }
  ]
}
```

I ten get this error:

> Unexpected token (Note that you need plugins to import files that are not JavaScript)

Looking at the source, I don't see anywhere that takes into account a project references build apart from the filter code I pasted at the start.

##### Versions

```text
typescript: 4.1.3
rollup: 2.37.0
rollup-plugin-typesscript2: 0.29.0
```

#### rollup.config.js

<details>
  <summary>`rollup.config.js`: </summary>

{
    input: inputFile,
    external: (id: string) => {
      if (id === 'babel-plugin-transform-async-to-promises/helpers') {
        return false;
      }

      return !id.startsWith('.') && !path.isAbsolute(id);
    },
    treeshake: {
      moduleSideEffects: false,
    },
    plugins: [
      eslint({
        fix: false,
        throwOnError: true,
        throwOnWarning: true,
        extensions: ['.ts', '.tsx', '.test.ts', '.test.tsx'],
        filterInclude: 'src/**',
        filterExclude: ['**/*.scss', '**/*.css', '**/*.md', '**/*.csv', 'dist/**', '**/*.json'],
        useEslintrc: true,
      }),
      resolve({
        mainFields: ['module', 'browser', 'main'],
        extensions: ['.mjs', '.cjs', '.js', '.ts', '.tsx', '.json', '.jsx'],
      }),
      json(),
      md(),
      postcss({
        extract: true,
        modules: false,
        autoModules: true,
        sourceMap: true,
        use: ['sass'],
        plugins: [
          url({
            url: 'inline',
          }),
        ],
      }),
      csv(),
      typescript({
        clean: true,
        typescript: require('typescript'),
        tsconfig: paths.tsConfig,
        abortOnError: true,
        tsconfigDefaults: {
          compilerOptions: {
            sourceMap: true,
            declaration: true,
            target: 'esnext',
            jsx: 'react-jsx',
          },
          useTsconfigDeclarationDir: true,
        },
        tsconfigOverride: {
          compilerOptions: {
            sourceMap: true,
            target: 'esnext',
          },
        },
      }),
      babel({
        exclude: /\/node_modules\/(core-js)\//,
        babelHelpers: 'runtime',
        ...babelConfig,
      } as RollupBabelInputPluginOptions),
      injectProcessEnv({
        NODE_ENV: env,
      }),
      svgo(),
      sourceMaps(),
      minify &&
        terser({
          compress: {
            keep_infinity: true,
            pure_getters: true,
            passes: 10,
          },
          ecma: 2016,
          toplevel: moduleFormat === 'cjs',
          format: {
            comments: 'all',
          },
        }),
    ].filter(Boolean),
  });
}

</details>

#### tsconfig.json

<details>
  <summary>`tsconfig.json`: </summary>

```json
{
  "extends": "../../tsconfig-base.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": "./src" 
  },
  "include": ["src/**/*.ts", "src/**/*.tsx"],
  "exclude": ["src/**/*.test.ts", "src/**/*.test.tsx"],
  "references": [
    { "path": "../util" }
  ]
}
```

</details>

#### plugin output with verbosity 3

<details>
  <summary>plugin output with verbosity 3: </summary>

</details>

**Comment by ezolenko** [comment]
Did you try not using `rootDir` in tsconfig at all?

**Comment by dagda1** [comment]
@ezolenko adding `"baseUrl" to the tsconfig fixed it:

```ts
{
  "extends": "../../tsconfig-base.json",
  "compilerOptions": {
    "baseUrl": ".",
    "outDir": "./dist",
    "rootDir": "./src",
  },
  "include": ["src/**/*.ts", "src/**/*.tsx"],
  "exclude": ["src/**/*.test.ts", "src/**/*.test.tsx"],
  "references": [
    { "path": "../util" }
  ]
}
```

I am both relieved and surprised!

**Comment by agilgur5** [comment]
After investigating in https://github.com/ezolenko/rollup-plugin-typescript2/issues/295#issuecomment-1139994454 a bit more, the reason project references failed here may have been due to an upstream issue with `@rollup/pluginutils`. 

https://github.com/rollup/plugins/pull/517 looks to fix absolute paths. And TS parses out `references`'s `path`s as absolute paths (it's possible that `baseUrl` changes that? don't think it does).

rpt2 v0.30+ [upgraded](https://github.com/ezolenko/rollup-plugin-typescript2/commit/c6f6e52933f520195364758cdf7fe8a371269611) `@rollup/pluginutils` to a version that includes that PR, so a newer rpt2 should fix this issue (OP was using rpt2 v0.29.0).
Can't quite confirm that without a repro though, and one isn't provided here, but thought I'd at least give my best guess hypothesis as of right now
